zerver/data_import/mattermost.py

```diff
@@ -0,0 +1,540 @@
+import os
+import logging
+import subprocess
+import ujson
+import re
+
+from typing import Any, Callable, Dict, List, Set
+
+from django.conf import settings
+from django.utils.timezone import now as timezone_now
+
+from zerver.models import Recipient
+from zerver.lib.utils import (
+    process_list_in_batches,
+)
+from zerver.data_import.import_util import ZerverFieldsT, build_zerver_realm, \
+    build_stream, build_realm, build_message, \
+    create_converted_data_files, make_subscriber_map, build_recipients, \
+    build_user_profile, build_stream_subscriptions, \
+    build_personal_subscriptions, build_user_message, SubscriberHandler
+
+from zerver.data_import.mattermost_user import UserHandler
+from zerver.data_import.sequencer import NEXT_ID, IdMapper
+
+def make_realm(realm_id: int, team: Dict[str, Any]) -> ZerverFieldsT:
+    # set correct realm details
+    NOW = float(timezone_now().timestamp())
+    domain_name = settings.EXTERNAL_HOST
+    realm_subdomain = team["name"]
+
+    zerver_realm = build_zerver_realm(realm_id, realm_subdomain, NOW, 'Mattermost')
+    realm = build_realm(zerver_realm, realm_id, domain_name)
+
+    # We may override these later.
+    realm['zerver_defaultstream'] = []
+
+    return realm
+
+def process_user(user_dict: Dict[str, Any], realm_id: int, team_name: str,
+                 user_handler: UserHandler, user_id_mapper: IdMapper) -> ZerverFieldsT:
+    def is_team_admin(user_dict: Dict[str, Any]) -> bool:
+        for team in user_dict["teams"]:
+            if team["name"] == team_name and "team_admin" in team["roles"]:
+                return True
+        return False
+
+    def get_full_name(user_dict: Dict[str, Any]) -> str:
+        full_name = "{} {}".format(user_dict["first_name"], user_dict["last_name"])
+        if full_name.strip():
+            return full_name
+        return user_dict['username']
+
+    full_name = get_full_name(user_dict)
+    delivery_email = user_dict['email']
+    is_realm_admin = is_team_admin(user_dict)
+    is_guest = False
+    short_name = user_dict['username']
+    date_joined = int(timezone_now().timestamp())
+    timezone = 'UTC'
+
+    if user_dict["is_mirror_dummy"]:
+        is_active = False
+        is_mirror_dummy = True
+        id = user_id_mapper.get(user_dict['username'])
+        email = 'mirror-{id}@example.com'.format(id=id)
+    else:
+        id = user_id_mapper.get(user_dict['username'])
+        is_active = True
+        is_mirror_dummy = False
+        email = user_dict['email']
+
+    return build_user_profile(
+        avatar_source='G',
```

Comment:
Does Mattermost have avatars we can import?

Reply:
The export data dump don't have any information regarding avatars. The only images present in the dump is that of custom emojis. 

```
~ $ ls
ENTERPRISE-EDITION-LICENSE.txt  bin                             data                            fonts                           plugins
NOTICE.txt                      client                          export_data.json                i18n                            prepackaged_plugins
README.md                       config                          exported_emoji                  logs                            templates
```

The mattermost server does  have a folder called `data` which contains images of custom emoji and users. 
```
~ $ ls data/
emoji  users
~ $ ls data/users/
8jfmpthpytbh8dyb4kqie7drpo  j7uoftdstpgyzf37rw6ubqdxjw  si6oizkfmbnqmf8pbphec15fpa
~ $ ls data/users/8jfmpthpytbh8dyb4kqie7drpo/
profile.png
```

But the file paths are hashed. The same emoji data is present as a folder in export dump called `exported_emoji` with additional information on what each file correspond to. If users can give us `user` data image folder as well we can probably write some script to map the hashed image paths to corresponding users.

Reply:
OK, we can just list this as something Mattermost doesn't export in the /help/ doc.